ql/src/java/org/apache/hadoop/hive/ql/optimizer/topnkey/TopNKeyPushdownProcessor.java

```diff
@@ -255,6 +260,88 @@ private void pushdownThroughLeftOuterJoin(TopNKeyOperator topNKey) throws Semant
     }
   }
 
+  private void pushdownInnerJoin(TopNKeyOperator topNKey, int fkJoinInputIndex, boolean nonFkSideIsFiltered) throws SemanticException {
+    TopNKeyDesc topNKeyDesc = topNKey.getConf();
+    CommonJoinOperator<? extends JoinDesc> join =
+            (CommonJoinOperator<? extends JoinDesc>) topNKey.getParentOperators().get(0);
+    List<Operator<? extends OperatorDesc>> joinInputs = join.getParentOperators();
+    ReduceSinkOperator fkJoinInput = (ReduceSinkOperator) joinInputs.get(fkJoinInputIndex);
+    if (nonFkSideIsFiltered) {
+      LOG.debug("Not pushing {} through {} as non FK side of the join is filtered", topNKey.getName(), join.getName());
+      return;
+    }
+    // Check column origins:
+    //  1. If all OrderBy columns are coming from the child (FK) table:
+    //    -> move TopNKeyOperator
+    //  2. If the first n OrderBy columns are coming from the child (FK) table:
+    //    -> copy TopNKeyOperator with the first n columns, and leave the original in place
+    int prefixLength = keyColumnPrefixLength(join, topNKey, fkJoinInputIndex, topNKey.getConf().getKeyColumns());
+    if (prefixLength == 0) {
+      LOG.debug("Not pushing {} through {} as common key column prefix length is 0", topNKey.getName(), join.getName());
+      return;
+    }
+    LOG.debug("Pushing a copy of {} through {} and {}",
+            topNKey.getName(), join.getName(), fkJoinInput.getName());
+    TopNKeyDesc newTopNKeyDesc = topNKeyDesc.withKeyColumns(prefixLength);
+    newTopNKeyDesc.setKeyColumns(remapColumns(join, fkJoinInput, newTopNKeyDesc.getKeyColumns()));
+    pushdown(copyDown(fkJoinInput, newTopNKeyDesc));
+    if (topNKeyDesc.getKeyColumns().size() == prefixLength) {
+      LOG.debug("Removing {} above {}", topNKey.getName(), join.getName());
+      join.removeChildAndAdoptItsChildren(topNKey);
+    }
+  }
+
+  private int keyColumnPrefixLength(CommonJoinOperator<? extends JoinDesc> join, TopNKeyOperator topNKeyOperator, int expectedTag, List<ExprNodeDesc> keyColumns) {
```

Comment:
nit. Some lines are very long. Please split in multiple lines.

Additionally, can we some comments to these utility methods?